File: ios/AI Crypto Analyzer/BinanceMarketService.swift
import Foundation

enum BinanceMarketError: LocalizedError {
    case invalidURL
    case requestFailed(String)
    case invalidPayload
    case unsupportedInterval(String)

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Unable to create Binance API URL."
        case .requestFailed(let message): return message
        case .invalidPayload: return "Binance returned market data in an unexpected format."
        case .unsupportedInterval(let interval): return "Binance does not support \(interval) candles on this endpoint."
        }
    }
}

enum BinanceMarketService {
    // Google Cloud (api-gcp) is prioritized because it supports true 1-second history natively!
    private static let restBaseURLs = [
        "https://api-gcp.binance.com",
        "https://api.binance.com",
        "https://api1.binance.com",
        "https://api.binance.us"
    ]

    struct CandleFetchResult {
        let candles: [Candle]
        let rawResponse: String
        let sourceURL: String
    }

    struct BookTickerFetchResult {
        let ticker: BookTicker
        let rawResponse: String
        let sourceURL: String
    }

    struct DepthFetchResult {
        let orderBook: OrderBookSnapshot
        let rawResponse: String
        let sourceURL: String
    }

    static func fetchCandles(
        symbol: String,
        timeframe: Timeframe,
        limit: Int = 100
    ) async throws -> [Candle] {
        try await fetchCandlesDetailed(symbol: symbol, timeframe: timeframe, limit: limit).candles
    }

    static func fetchCandlesDetailed(
        symbol: String,
        timeframe: Timeframe,
        limit: Int = 100
    ) async throws -> CandleFetchResult {
        var lastError: Error?

        // 1. Try to fetch standard klines first (even for 1s, since GCP supports it)
        for baseURL in restBaseURLs {
            do {
                let result = try await fetchKlines(
                    baseURL: baseURL,
                    symbol: symbol,
                    timeframe: timeframe,
                    limit: limit
                )
                if !result.candles.isEmpty { return result }
            } catch {
                lastError = error
                continue
            }
        }
        
        // 2. If Klines completely fail (like on Binance.US), manually build 1s from recent trades
        if timeframe == .oneSecond {
            return try await fetchOneSecondFromTrades(symbol: symbol, limit: limit)
        }

        throw lastError ?? BinanceMarketError.invalidPayload
    }

    static func fetchBookTickerDetailed(symbol: String) async throws -> BookTickerFetchResult {
        var lastError: Error?

        for baseURL in restBaseURLs {
            do {
                return try await fetchBookTicker(baseURL: baseURL, symbol: symbol)
            } catch {
                lastError = error
            }
        }

        throw lastError ?? BinanceMarketError.invalidPayload
    }

    static func fetchDepthDetailed(symbol: String, limit: Int = 20) async throws -> DepthFetchResult {
        var lastError: Error?

        for baseURL in restBaseURLs {
            do {
                return try await fetchDepth(baseURL: baseURL, symbol: symbol, limit: limit)
            } catch {
                lastError = error
            }
        }

        throw lastError ?? BinanceMarketError.invalidPayload
    }

    private static func fetchOneSecondFromTrades(symbol: String, limit: Int) async throws -> CandleFetchResult {
        let tradeLimit = max(limit * 4, 1000)
        var lastError: Error?

        for baseURL in restBaseURLs {
            do {
                let tradeResult = try await fetchRecentTrades(
                    baseURL: baseURL,
                    symbol: symbol,
                    limit: tradeLimit
                )
                let candles = aggregateTradesIntoSecondCandles(tradeResult.trades).suffix(limit)
                if !candles.isEmpty {
                    return CandleFetchResult(
                        candles: Array(candles),
                        rawResponse: tradeResult.rawResponse,
                        sourceURL: tradeResult.sourceURL
                    )
                }
            } catch {
                lastError = error
            }
        }
        throw lastError ?? BinanceMarketError.invalidPayload
    }

    private static func fetchKlines(
        baseURL: String,
        symbol: String,
        timeframe: Timeframe,
        limit: Int
    ) async throws -> CandleFetchResult {
        var components = URLComponents(string: "\(baseURL)/api/v3/klines")
        components?.queryItems = [
            URLQueryItem(name: "symbol", value: symbol),
            URLQueryItem(name: "interval", value: timeframe.rawValue),
            URLQueryItem(name: "limit", value: String(limit))
        ]

        guard let url = components?.url else { throw BinanceMarketError.invalidURL }
        
        var request = URLRequest(url: url)
        request.timeoutInterval = 4
        request.cachePolicy = .reloadIgnoringLocalCacheData 

        let (data, response) = try await URLSession.shared.data(for: request)
        if let response = response as? HTTPURLResponse, !(200...299).contains(response.statusCode) {
            let body = String(data: data, encoding: .utf8) ?? "No response body."
            throw BinanceMarketError.requestFailed("Binance HTTP \(response.statusCode): \(body)")
        }

        let payload = try JSONSerialization.jsonObject(with: data)
        guard let rows = payload as? [[Any]] else { throw BinanceMarketError.invalidPayload }

        let candles: [Candle] = rows.compactMap { row -> Candle? in
            guard row.count >= 6,
                  let openTime = doubleValue(row[0]),
                  let open = doubleValue(row[1]),
                  let high = doubleValue(row[2]),
                  let low = doubleValue(row[3]),
                  let close = doubleValue(row[4]),
                  let volume = doubleValue(row[5]) else {
                return nil
            }
            return Candle(
                openTime: Date(timeIntervalSince1970: openTime / 1000),
                open: open, high: high, low: low, close: close, volume: volume
            )
        }

        return CandleFetchResult(
            candles: candles,
            rawResponse: String(data: data, encoding: .utf8) ?? "Unable to decode raw response.",
            sourceURL: url.absoluteString
        )
    }

    private static func fetchBookTicker(baseURL: String, symbol: String) async throws -> BookTickerFetchResult {
        var components = URLComponents(string: "\(baseURL)/api/v3/ticker/bookTicker")
        components?.queryItems = [
            URLQueryItem(name: "symbol", value: symbol)
        ]

        guard let url = components?.url else { throw BinanceMarketError.invalidURL }
        let data = try await fetchData(from: url)
        let response = try JSONDecoder().decode(BinanceBookTickerResponse.self, from: data)

        guard let bidPrice = Double(response.bidPrice),
              let bidQuantity = Double(response.bidQty),
              let askPrice = Double(response.askPrice),
              let askQuantity = Double(response.askQty) else {
            throw BinanceMarketError.invalidPayload
        }

        return BookTickerFetchResult(
            ticker: BookTicker(
                symbol: response.symbol,
                bidPrice: bidPrice,
                bidQuantity: bidQuantity,
                askPrice: askPrice,
                askQuantity: askQuantity
            ),
            rawResponse: String(data: data, encoding: .utf8) ?? "Unable to decode raw response.",
            sourceURL: url.absoluteString
        )
    }

    private static func fetchDepth(baseURL: String, symbol: String, limit: Int) async throws -> DepthFetchResult {
        var components = URLComponents(string: "\(baseURL)/api/v3/depth")
        components?.queryItems = [
            URLQueryItem(name: "symbol", value: symbol),
            URLQueryItem(name: "limit", value: String(limit))
        ]

        guard let url = components?.url else { throw BinanceMarketError.invalidURL }
        let data = try await fetchData(from: url)
        let response = try JSONDecoder().decode(BinanceDepthResponse.self, from: data)

        return DepthFetchResult(
            orderBook: OrderBookSnapshot(
                lastUpdateId: response.lastUpdateId,
                bids: parseOrderBookLevels(response.bids),
                asks: parseOrderBookLevels(response.asks)
            ),
            rawResponse: String(data: data, encoding: .utf8) ?? "Unable to decode raw response.",
            sourceURL: url.absoluteString
        )
    }

    private static func fetchData(from url: URL) async throws -> Data {
        var request = URLRequest(url: url)
        request.timeoutInterval = 4
        request.cachePolicy = .reloadIgnoringLocalCacheData

        let (data, response) = try await URLSession.shared.data(for: request)
        if let response = response as? HTTPURLResponse, !(200...299).contains(response.statusCode) {
            let body = String(data: data, encoding: .utf8) ?? "No response body."
            throw BinanceMarketError.requestFailed("Binance HTTP \(response.statusCode): \(body)")
        }
        return data
    }

    private static func parseOrderBookLevels(_ rows: [[String]]) -> [OrderBookLevel] {
        rows.compactMap { row in
            guard row.count >= 2,
                  let price = Double(row[0]),
                  let quantity = Double(row[1]) else {
                return nil
            }
            return OrderBookLevel(price: price, quantity: quantity)
        }
    }

    private struct RecentTradeFetchResult {
        let trades: [BinanceRecentTrade]
        let rawResponse: String
        let sourceURL: String
    }

    private static func fetchRecentTrades(
        baseURL: String,
        symbol: String,
        limit: Int
    ) async throws -> RecentTradeFetchResult {
        var components = URLComponents(string: "\(baseURL)/api/v3/trades")
        components?.queryItems = [
            URLQueryItem(name: "symbol", value: symbol),
            URLQueryItem(name: "limit", value: String(limit))
        ]

        guard let url = components?.url else { throw BinanceMarketError.invalidURL }
        
        var request = URLRequest(url: url)
        request.timeoutInterval = 4
        request.cachePolicy = .reloadIgnoringLocalCacheData 

        let (data, response) = try await URLSession.shared.data(for: request)
        if let response = response as? HTTPURLResponse, !(200...299).contains(response.statusCode) {
            let body = String(data: data, encoding: .utf8) ?? "No response body."
            throw BinanceMarketError.requestFailed("Binance HTTP \(response.statusCode): \(body)")
        }

        return RecentTradeFetchResult(
            trades: try JSONDecoder().decode([BinanceRecentTrade].self, from: data),
            rawResponse: String(data: data, encoding: .utf8) ?? "Unable to decode raw response.",
            sourceURL: url.absoluteString
        )
    }

    private static func aggregateTradesIntoSecondCandles(_ trades: [BinanceRecentTrade]) -> [Candle] {
        let sortedTrades = trades.sorted { $0.resolvedTime < $1.resolvedTime }
        var candles: [Candle] = []

        for trade in sortedTrades {
            let secondStart = Date(timeIntervalSince1970: floor(Double(trade.resolvedTime) / 1000))
            let price = Double(trade.resolvedPrice) ?? 0
            let volume = Double(trade.resolvedQuantity) ?? 0
            guard price > 0 else { continue }

            if let lastIndex = candles.indices.last, candles[lastIndex].openTime == secondStart {
                let last = candles[lastIndex]
                candles[lastIndex] = Candle(
                    openTime: last.openTime, open: last.open, high: max(last.high, price),
                    low: min(last.low, price), close: price, volume: last.volume + volume
                )
            } else {
                let openPrice = candles.last?.close ?? price
                candles.append(Candle(
                    openTime: secondStart, open: openPrice, high: price,
                    low: price, close: price, volume: volume
                ))
            }
        }
        return candles
    }

    private static func doubleValue(_ value: Any) -> Double? {
        if let double = value as? Double { return double }
        if let string = value as? String { return Double(string) }
        if let number = value as? NSNumber { return number.doubleValue }
        return nil
    }
}

private struct BinanceBookTickerResponse: Decodable {
    let symbol: String
    let bidPrice: String
    let bidQty: String
    let askPrice: String
    let askQty: String
}

private struct BinanceDepthResponse: Decodable {
    let lastUpdateId: Int
    let bids: [[String]]
    let asks: [[String]]
}

// MARK: - WEBSOCKETS

struct TradeTick: Equatable {
    let time: Date
    let price: Double
    let quantity: Double
}

final class BinanceWebSocketService {
    private var task: URLSessionWebSocketTask?
    private var pingTimer: Timer?
    private var lastMessageTime: Date = .distantPast
    private var isDisconnected = false
    
    private let wssBaseURLs = [
        "wss://stream1.binance.com:443",     // Spot GCP Cluster (Best for spot)
        "wss://stream.binance.com:9443",     // Spot Main
        "wss://fstream.binance.com:443",     // Futures (Fallback)
        "wss://stream.binance.us:9443"       // US Spot (Never Blocked)
    ]
    
    private var urlIndex = 0
    private var symbol = ""
    private var timeframe = Timeframe.fiveMinutes
    private var onCandle: ((Candle) -> Void)?
    private var onError: ((String) -> Void)?

    func connect(
        symbol: String,
        timeframe: Timeframe,
        onCandle: @escaping @MainActor (Candle) -> Void,
        onError: @escaping @MainActor (String) -> Void
    ) {
        disconnect()
        self.symbol = symbol
        self.timeframe = timeframe
        self.onCandle = onCandle
        self.onError = onError
        self.urlIndex = 0
        self.isDisconnected = false
        connectToCurrentURL()
        startPing()
    }
    
    private func connectToCurrentURL() {
        let stream = "\(symbol.lowercased())@kline_\(timeframe.rawValue)"
        let baseURL = wssBaseURLs[urlIndex]
        
        guard let url = URL(string: "\(baseURL)/ws/\(stream)") else { return }

        task = URLSession.shared.webSocketTask(with: url)
        task?.resume()
        lastMessageTime = Date()
        receive()
    }

    func disconnect() {
        isDisconnected = true
        pingTimer?.invalidate()
        pingTimer = nil
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
    }
    
    // MARK: - Keepalive Ping
    private func startPing() {
        pingTimer?.invalidate()
        pingTimer = Timer.scheduledTimer(withTimeInterval: 20, repeats: true) { [weak self] _ in
            self?.sendPing()
            self?.checkConnectionHealth()
        }
    }
    
    private func sendPing() {
        task?.sendPing { [weak self] error in
            if let error = error {
                print("WebSocket ping failed: \(error.localizedDescription)")
                self?.handleDisconnect(reason: "Ping failed")
            }
        }
    }
    
    private func checkConnectionHealth() {
        let staleSeconds = Date().timeIntervalSince(lastMessageTime)
        // If no message received for 60 seconds, connection is likely dead
        if staleSeconds > 60 && !isDisconnected {
            handleDisconnect(reason: "No data for \(Int(staleSeconds))s")
        }
    }
    
    private func handleDisconnect(reason: String) {
        Task { @MainActor in
            self.onError?("Connection lost: \(reason). Reconnecting...")
        }
        
        // Try next server
        urlIndex += 1
        if urlIndex < wssBaseURLs.count {
            connectToCurrentURL()
        } else {
            urlIndex = 0
            DispatchQueue.global().asyncAfter(deadline: .now() + 2) { [weak self] in
                self?.connectToCurrentURL()
            }
        }
    }

    private func receive() {
        task?.receive { [weak self] result in
            guard let self = self, !self.isDisconnected else { return }

            switch result {
            case .success(let message):
                self.lastMessageTime = Date()
                if let candle = Self.decode(message: message) {
                    Task { @MainActor in self.onCandle?(candle) }
                }
                self.receive()
                
            case .failure(_):
                self.handleDisconnect(reason: "Receive error")
            }
        }
    }

    private static func decode(message: URLSessionWebSocketTask.Message) -> Candle? {
        let data: Data?
        switch message {
        case .data(let payload): data = payload
        case .string(let payload): data = payload.data(using: .utf8)
        @unknown default: data = nil
        }
        guard let data else { return nil }
        guard let msg = try? JSONDecoder().decode(BinanceKlineMessage.self, from: data) else { return nil }
        return msg.candle
    }
}

final class BinanceTradeWebSocketService {
    private var task: URLSessionWebSocketTask?
    private var pingTimer: Timer?
    private var lastMessageTime: Date = .distantPast
    private var isDisconnected = false
    
    private let wssBaseURLs = [
        "wss://stream1.binance.com:443",     // Spot GCP Cluster (Best for spot)
        "wss://stream.binance.com:9443",     // Spot Main
        "wss://fstream.binance.com:443",     // Futures (Fallback)
        "wss://stream.binance.us:9443"       // US Spot (Never Blocked)
    ]
    
    private var urlIndex = 0
    private var symbol = ""
    private var onTrade: ((TradeTick) -> Void)?
    private var onError: ((String) -> Void)?

    func connect(
        symbol: String,
        onTrade: @escaping @MainActor (TradeTick) -> Void,
        onError: @escaping @MainActor (String) -> Void
    ) {
        disconnect()
        self.symbol = symbol
        self.onTrade = onTrade
        self.onError = onError
        self.urlIndex = 0
        self.isDisconnected = false
        connectToCurrentURL()
        startPing()
    }
    
    private func connectToCurrentURL() {
        let stream = "\(symbol.lowercased())@aggTrade"
        let baseURL = wssBaseURLs[urlIndex]
        
        guard let url = URL(string: "\(baseURL)/ws/\(stream)") else { return }

        task = URLSession.shared.webSocketTask(with: url)
        task?.resume()
        lastMessageTime = Date()
        receive()
    }

    func disconnect() {
        isDisconnected = true
        pingTimer?.invalidate()
        pingTimer = nil
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
    }
    
    // MARK: - Keepalive Ping
    private func startPing() {
        pingTimer?.invalidate()
        pingTimer = Timer.scheduledTimer(withTimeInterval: 20, repeats: true) { [weak self] _ in
            self?.sendPing()
            self?.checkConnectionHealth()
        }
    }
    
    private func sendPing() {
        task?.sendPing { [weak self] error in
            if let error = error {
                print("Trade WebSocket ping failed: \(error.localizedDescription)")
                self?.handleDisconnect(reason: "Ping failed")
            }
        }
    }
    
    private func checkConnectionHealth() {
        let staleSeconds = Date().timeIntervalSince(lastMessageTime)
        // Trade stream should get messages very frequently (BTC trades every second)
        // If no message for 30 seconds, connection is likely dead
        if staleSeconds > 30 && !isDisconnected {
            handleDisconnect(reason: "No trades for \(Int(staleSeconds))s")
        }
    }
    
    private func handleDisconnect(reason: String) {
        Task { @MainActor in
            self.onError?("Connection lost: \(reason). Reconnecting...")
        }
        
        urlIndex += 1
        if urlIndex < wssBaseURLs.count {
            connectToCurrentURL()
        } else {
            urlIndex = 0
            DispatchQueue.global().asyncAfter(deadline: .now() + 2) { [weak self] in
                self?.connectToCurrentURL()
            }
        }
    }

    private func receive() {
        task?.receive { [weak self] result in
            guard let self = self, !self.isDisconnected else { return }

            switch result {
            case .success(let message):
                self.lastMessageTime = Date()
                if let trade = Self.decode(message: message) {
                    Task { @MainActor in self.onTrade?(trade) }
                }
                self.receive()
                
            case .failure(_):
                self.handleDisconnect(reason: "Receive error")
            }
        }
    }

    private static func decode(message: URLSessionWebSocketTask.Message) -> TradeTick? {
        let data: Data?
        switch message {
        case .data(let payload): data = payload
        case .string(let payload): data = payload.data(using: .utf8)
        @unknown default: data = nil
        }
        guard let data else { return nil }
        guard let msg = try? JSONDecoder().decode(BinanceTradeMessage.self, from: data) else { return nil }
        return msg.trade
    }
}

// MARK: - Bulletproof Universal Decoder Models

private struct BinanceRecentTrade: Decodable {
    let price: String?
    let p: String?
    let qty: String?
    let q: String?
    let time: Int?
    let T: Int?
    
    var resolvedPrice: String { price ?? p ?? "0" }
    var resolvedQuantity: String { qty ?? q ?? "0" }
    var resolvedTime: Int { time ?? T ?? 0 }
}

private struct BinanceTradeMessage: Decodable {
    let p: String? // Price
    let q: String? // Quantity
    let T: Int?    // Trade time
    let E: Int?    // Event time

    var trade: TradeTick? {
        guard let priceStr = p, let price = Double(priceStr),
              let qtyStr = q, let quantity = Double(qtyStr) else { return nil }
        
        let timestamp = T ?? E ?? Int(Date().timeIntervalSince1970 * 1000)
        return TradeTick(
            time: Date(timeIntervalSince1970: Double(timestamp) / 1000),
            price: price,
            quantity: quantity
        )
    }
}

private struct BinanceKlineMessage: Decodable {
    let k: BinanceKline?

    var candle: Candle? {
        guard let kline = k,
              let t = kline.t,
              let o = Double(kline.o ?? "0"),
              let h = Double(kline.h ?? "0"),
              let l = Double(kline.l ?? "0"),
              let c = Double(kline.c ?? "0"),
              let v = Double(kline.v ?? "0") else { return nil }
        
        return Candle(
            openTime: Date(timeIntervalSince1970: Double(t) / 1000),
            open: o, high: h, low: l, close: c, volume: v
        )
    }
}

private struct BinanceKline: Decodable {
    let t: Int?
    let o: String?
    let h: String?
    let l: String?
    let c: String?
    let v: String?
}

File: web/src/services/BinanceMarketService.ts
// ============================================================
// Binance Market Service — Ported from BinanceMarketService.swift
// ============================================================

import type { Candle, BookTicker, OrderBookSnapshot, TradeTick } from '../engine/types';

const REST_BASE_URLS = [
  'https://api.binance.com',
  'https://api1.binance.com',
  'https://api2.binance.com',
  'https://api.binance.us',
];

const WSS_BASE_URLS = [
  'wss://stream.binance.com:9443',
  'wss://stream1.binance.com:443',
  'wss://stream.binance.us:9443',
];

// ---- REST API ----

// CoinGecko coin ID mapping
const COINGECKO_COIN_IDS: Record<string, string> = {
  BTCUSDT: 'bitcoin',
  ETHUSDT: 'ethereum',
  SOLUSDT: 'solana',
  BNBUSDT: 'binancecoin',
  XRPUSDT: 'ripple',
  DOGEUSDT: 'dogecoin',
};
const COINGECKO_TIMEFRAMES: Record<string, { days: number | string; granularity: string }> = {
  '1m': { days: '0.042', granularity: '' },    // ~1 hour
  '5m': { days: '1', granularity: '' },          // CoinGecko doesn't have 5m, use 1d
  '15m': { days: '1', granularity: '' },
  '1h': { days: '7', granularity: '' },
  '4h': { days: '30', granularity: '' },
  '1d': { days: '90', granularity: 'daily' },
  '1w': { days: '365', granularity: 'weekly' },
  '1M': { days: 'max', granularity: 'monthly' },
};

export async function fetchCandles(symbol: string, timeframe: string, limit: number = 100): Promise<Candle[]> {
  // Try Binance first
  for (const baseURL of REST_BASE_URLS) {
    try {
      const url = `${baseURL}/api/v3/klines?symbol=${symbol}&interval=${timeframe}&limit=${limit}`;
      const res = await fetch(url);
      if (!res.ok) continue;
      const data = await res.json();
      if (Array.isArray(data) && data.length > 0) {
        return data.map((k: any[]) => ({
          openTime: k[0],
          open: parseFloat(k[1]),
          high: parseFloat(k[2]),
          low: parseFloat(k[3]),
          close: parseFloat(k[4]),
          volume: parseFloat(k[5]),
          isClosed: true, // REST API candles are always closed
        }));
      }
    } catch { continue; }
  }

  // Fallback to CoinGecko OHLC
  try {
    const cg = COINGECKO_TIMEFRAMES[timeframe];
    const days = cg ? cg.days : '1';
    const coinId = COINGECKO_COIN_IDS[symbol] ?? 'bitcoin';
    const url = `https://api.coingecko.com/api/v3/coins/${coinId}/ohlc?vs_currency=usd&days=${days}`;
    const res = await fetch(url);
    if (res.ok) {
      const data = await res.json();
      if (Array.isArray(data) && data.length > 0) {
        return data.map((k: number[]) => ({
          openTime: k[0],
          open: k[1],
          high: k[2],
          low: k[3],
          close: k[4],
          volume: 0, // CoinGecko OHLC doesn't include volume
        }));
      }
    }
  } catch {
    // CoinGecko failed too
  }

  return [];
}

export async function fetchBookTicker(symbol: string): Promise<BookTicker | null> {
  for (const baseURL of REST_BASE_URLS) {
    try {
      const url = `${baseURL}/api/v3/ticker/bookTicker?symbol=${symbol}`;
      const res = await fetch(url);
      if (!res.ok) continue;
      const data = await res.json();
      return {
        symbol: data.symbol,
        bidPrice: parseFloat(data.bidPrice),
        bidQuantity: parseFloat(data.bidQty),
        askPrice: parseFloat(data.askPrice),
        askQuantity: parseFloat(data.askQty),
      };
    } catch { continue; }
  }
  return null;
}

export async function fetchDepth(symbol: string, limit: number = 10): Promise<OrderBookSnapshot | null> {
  for (const baseURL of REST_BASE_URLS) {
    try {
      const url = `${baseURL}/api/v3/depth?symbol=${symbol}&limit=${limit}`;
      const res = await fetch(url);
      if (!res.ok) continue;
      const data = await res.json();
      return {
        lastUpdateId: data.lastUpdateId,
        bids: data.bids.slice(0, limit).map((b: string[]) => ({ price: parseFloat(b[0]), quantity: parseFloat(b[1]) })),
        asks: data.asks.slice(0, limit).map((a: string[]) => ({ price: parseFloat(a[0]), quantity: parseFloat(a[1]) })),
      };
    } catch { continue; }
  }
  return null;
}

// ---- WebSocket Services ----

export class BinanceKlineWebSocket {
  private ws: WebSocket | null = null;
  private pingInterval: ReturnType<typeof setInterval> | null = null;
  private healthInterval: ReturnType<typeof setInterval> | null = null;
  private lastMessageTime = 0;
  private urlIndex = 0;
  private symbol = '';
  private timeframe = '5m';
  private onCandle: (candle: Candle) => void = () => {};
  private onError: (msg: string) => void = () => {};
  private isDisconnected = false;

  connect(symbol: string, timeframe: string, onCandle: (c: Candle) => void, onError: (msg: string) => void) {
    this.disconnect();
    this.symbol = symbol;
    this.timeframe = timeframe;
    this.onCandle = onCandle;
    this.onError = onError;
    this.urlIndex = 0;
    this.isDisconnected = false;
    this.connectToCurrentURL();
    this.startPing();
  }

  private connectToCurrentURL() {
    const stream = `${this.symbol.toLowerCase()}@kline_${this.timeframe}`;
    const baseURL = WSS_BASE_URLS[this.urlIndex];
    const url = `${baseURL}/ws/${stream}`;

    this.ws = new WebSocket(url);
    this.lastMessageTime = Date.now();

    this.ws.onmessage = (event) => {
      this.lastMessageTime = Date.now();
      try {
        const msg = JSON.parse(event.data);
        if (msg.k) {
          const k = msg.k;
          this.onCandle({
            openTime: k.t,
            open: parseFloat(k.o),
            high: parseFloat(k.h),
            low: parseFloat(k.l),
            close: parseFloat(k.c),
            volume: parseFloat(k.v),
            isClosed: !!k.x, // Binance k.x = true when candle is closed
          });
        }
      } catch {}
    };

    this.ws.onerror = () => {
      this.handleDisconnect('WebSocket error');
    };

    this.ws.onclose = () => {
      if (!this.isDisconnected) {
        this.handleDisconnect('Connection closed');
      }
    };
  }

  disconnect() {
    this.isDisconnected = true;
    if (this.pingInterval) clearInterval(this.pingInterval);
    if (this.healthInterval) clearInterval(this.healthInterval);
    this.pingInterval = null;
    this.healthInterval = null;
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  private startPing() {
    if (this.pingInterval) clearInterval(this.pingInterval);
    this.pingInterval = setInterval(() => {
      if (this.ws?.readyState === WebSocket.OPEN) {
        this.ws.send(JSON.stringify({ method: 'ping' }));
      }
    }, 20000);

    if (this.healthInterval) clearInterval(this.healthInterval);
    this.healthInterval = setInterval(() => {
      const stale = (Date.now() - this.lastMessageTime) / 1000;
      if (stale > 60 && !this.isDisconnected) {
        this.handleDisconnect(`No data for ${Math.floor(stale)}s`);
      }
    }, 10000);
  }

  private handleDisconnect(reason: string) {
    this.onError(`Connection lost: ${reason}. Reconnecting...`);
    this.urlIndex++;
    if (this.urlIndex < WSS_BASE_URLS.length) {
      this.connectToCurrentURL();
    } else {
      this.urlIndex = 0;
      setTimeout(() => this.connectToCurrentURL(), 2000);
    }
  }
}

export class BinanceTradeWebSocket {
  private ws: WebSocket | null = null;
  private pingInterval: ReturnType<typeof setInterval> | null = null;
  private healthInterval: ReturnType<typeof setInterval> | null = null;
  private lastMessageTime = 0;
  private symbol = '';
  private onTrade: (t: TradeTick) => void = () => {};
  private onError: (msg: string) => void = () => {};
  private isDisconnected = false;

  connect(symbol: string, onTrade: (t: TradeTick) => void, onError: (msg: string) => void) {
    this.disconnect();
    this.symbol = symbol;
    this.onTrade = onTrade;
    this.onError = onError;
    this.isDisconnected = false;
    this.connectToCurrentURL();
    this.startPing();
  }

  private connectToCurrentURL() {
    // Use Binance Futures BookTicker for lower latency
    const stream = `${this.symbol.toLowerCase()}@bookTicker`;
    const url = `wss://fstream.binance.com/ws/${stream}`;

    this.ws = new WebSocket(url);
    this.lastMessageTime = Date.now();

    this.ws.onmessage = (event) => {
      this.lastMessageTime = Date.now();
      try {
        const msg = JSON.parse(event.data);
        // Futures bookTicker: { b: bestBid, B: bestBidQty, a: bestAsk, A: bestAskQty, ... }
        const price = msg.b ? parseFloat(msg.b) : (msg.a ? parseFloat(msg.a) : null);
        if (price) {
          this.onTrade({
            price,
            quantity: msg.B ? parseFloat(msg.B) : 0,
            time: msg.E || msg.T || Date.now(),
          });
        }
      } catch {}
    };

    this.ws.onerror = () => {
      this.handleDisconnect('WebSocket error');
    };

    this.ws.onclose = () => {
      if (!this.isDisconnected) {
        this.handleDisconnect('Connection closed');
      }
    };
  }

  disconnect() {
    this.isDisconnected = true;
    if (this.pingInterval) clearInterval(this.pingInterval);
    if (this.healthInterval) clearInterval(this.healthInterval);
    this.pingInterval = null;
    this.healthInterval = null;
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  private startPing() {
    if (this.pingInterval) clearInterval(this.pingInterval);
    this.pingInterval = setInterval(() => {
      if (this.ws?.readyState === WebSocket.OPEN) {
        this.ws.send(JSON.stringify({ method: 'ping' }));
      }
    }, 20000);

    if (this.healthInterval) clearInterval(this.healthInterval);
    this.healthInterval = setInterval(() => {
      const stale = (Date.now() - this.lastMessageTime) / 1000;
      if (stale > 30 && !this.isDisconnected) {
        this.handleDisconnect(`No trades for ${Math.floor(stale)}s`);
      }
    }, 10000);
  }

  private handleDisconnect(reason: string) {
    this.onError(`Connection lost: ${reason}. Reconnecting...`);
    // Reconnect to the same Futures BookTicker URL after a delay
    setTimeout(() => this.connectToCurrentURL(), 2000);
  }
}
